Add `version_epoch` batch type and use it for offset retention feature

dotnwat opened the pull request:
* Introduces a new `version_fence` batch type that can be used to establish an epoch within a partition for the purposes of reasoning about the age of batches with respect to the activation of features and upgrades.
* Uses the `version_fence` within consumer group partitions to identify committed offsets that were written prior to upgrading to v23.
* Disables offset retention for committed offsets written prior to v23 as a conservative policy that addresses the omission of key metadata in prior committed offsets that make evaluating the age of said offsets ambiguous within the context of the retention policy.

Fixes: https://github.com/redpanda-data/redpanda/issues/8348

<!--

See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md##pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED.

Describe, in plain language, the motivation behind the change (bug fix,
feature, improvement) in this PR and how the included commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.

  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.

  Backport of PR #PR-NUMBER

-->

## Backports Required

<!--

Checking at least one of the checkboxes is REQUIRED if this PR is not a backport.

-->

- [x] none - not a bug fix
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v22.3.x
- [ ] v22.2.x
- [ ] v22.1.x

## UX Changes

<!--

Content in this section is OPTIONAL.

Describe, in plain language, how this PR affects an end-user. Explain
topic flags, configuration flags, command line flags, deprecation
policies, etc. that are added or modified. Don't ship user breaking
changes. Ask the @redpanda-data/product team if you need help with user
visible changes.

-->

## Release Notes

<!--

Adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

  ### Bug Fixes

  * Short description of the bug fix if this is a PR to `dev` branch.
-->

### Features

Offset retention will be enabled in v23 which enables automatic deletion of old committed offsets. The policy is the same as Kafka. Behavior of retention differs for offsets committed prior to upgrading.

* If an offset was committed prior to upgrade then it is not automatically reclaimed under any policy. To remove such offsets use the Kafka Delete Offset API.
* If an offset was committed prior to upgrade, and later re-committed / written after upgrade then behavior is identical to a committed offset that was initially written after the upgrade. Namely, normal offset retention policy is active.
* Legacy systems must explicitly enable offset retention following an upgrade. It is important to note the behavior of the system in this case. Offsets committed after upgrade, but before offset retention is enabled, will behave as if retention had always been enabled. For example, if an offset were committed the same day as an upgrade with default retention time of 7 days, and then legacy support was enabled 5 days later, the offset would be reclaimed 2 days after.

mmaslankaprv commented on src/v/features/feature_table.cc:
```diff
@@ -423,6 +423,20 @@ void feature_table::testing_activate_all() {
     on_update();
 }
 
+feature_table::version_fence
```
should commit message start from `features:` ?

dotnwat replied:
i went back and forth on it. changed to `features:`

dotnwat commented on src/v/features/feature_table.cc:
```diff
@@ -423,6 +423,20 @@ void feature_table::testing_activate_all() {
 
+feature_table::version_fence
+feature_table::decode_version_fence(model::record_batch batch) {
+    auto records = batch.copy_records();
```
did you mean something else, because `record_batch::release_records` doesn't exist. copy records, despite the name, doesn't really copy records that already exist, it decodes records from the internal iobuf into a vector as a convenience over using the for_each_record pattern with a lambda.

dotnwat commented on src/v/features/feature_table.h:
```diff
@@ -345,6 +346,38 @@ class feature_table {
     void bootstrap_active_version(cluster::cluster_version);
 
+    /*
+     *
```
fixed

dotnwat commented on src/v/features/feature_table.h:
```diff
@@ -345,6 +346,38 @@ class feature_table {
+    static constexpr std::string_view version_fence_batch_key = "state";
+    static version_fence decode_version_fence(model::record_batch batch);
+
+    model::record_batch encode_version_fence() const {
```
great idea, i wasn't too happy with the interface either. fixed.

dotnwat commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -657,6 +658,62 @@ ss::future<> group_manager::recover_partition(
+            if (result) {
+                vlog(
+                  klog.info,
+                  "Wrote offset retention feature fence for partition {}",
```
Oh I really like this mental model regarding nouns that user is familiar with for INFO level guidance.

dotnwat commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -657,6 +658,62 @@ ss::future<> group_manager::recover_partition(
+                  result.error());
+            }
+        } catch (...) {
+            vlog(
```
1. Consumer groups will continue to work as expected even if writing the fence fails
2. Offset retention will even function correctly if writing the fence fails
3. When the fence fails to be written, the scenario of interest is that committed offsets that slip in before the fence does eventually get written will be misclassified as being from before the upgrade.

In the latest force push, the window of vulnerability for (3) is shrunk even more because we put the fence writing in a background retry loop. the backgrounding was required in any case because I forgot to take into account that I couldn't await_feature in this recovery path without making consumer groups unavailable during a partition upgrade.

bharathv reviewed (commented):
lgtm, just nit-picks.

dotnwat commented on src/v/kafka/server/group.h:
```diff
@@ -157,6 +157,13 @@ class group final : public ss::enable_lw_shared_from_this<group> {
+         * persistent retention metadata we act conservatively and skip
+         * automatic reclaim. offset delete api can be used to remove them.
+         */
+        bool non_reclaimable{false};
```
yeh i think the double negative is a bit awkward. the intention was that in the place where it was used to implement the policy it ended up avoiding the `!` negation.

```
if (non_reclaimable) {
  return;
}
```

dotnwat commented:
Force push changes:

1. Rebased to fix feature table conflict
2. Updated comments
3. Feature fence encoding takes min version predicate to reinforce common usage
4. Fixed shutdown hang in consumer groups where partition abort was not requested
5. await_feature in background for version fence writing to fix unavailability of consumer groups in mixed version clusters
6. Retry version fence appends
7. Fixed error level when exception occurs during fence replicate

andrwng commented on src/v/features/feature_table.cc:
```diff
@@ -446,6 +446,20 @@ void feature_table::testing_activate_all() {
+feature_table::decode_version_fence(model::record_batch batch) {
+    auto records = batch.copy_records();
+    if (records.empty()) {
+        throw std::runtime_error("");
```
nit: could you make these throw messages clearer so we can at least identify what's wrong with the batch in case there are serialization issues?

andrwng commented on src/v/features/feature_table.h:
```diff
@@ -363,6 +364,55 @@ class feature_table {
     void abort_for_tests() { _as.request_abort(); }
 
+    /*
+     * The verison fence is the structure encoded into a verison fence batch
```
nit: s/verison/version

andrwng commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -746,6 +762,75 @@ ss::future<> group_manager::do_recover_group(
+    // how long to delay retrying a fence write if an error occurs
+    constexpr auto fence_write_retry_delay = 10s;
+
+    co_await _feature_table.local().await_feature(
+      features::feature::group_offset_retention, p->as);
```
nit: if you rebase you could try using the `await_feature_then()` method to encapsulate some of the failure handling (may not be worth it, considering this uses the partition's abort source rather than the feature table's)

andrwng commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -746,6 +762,75 @@ ss::future<> group_manager::do_recover_group(
+            break;
+        }
+
+        auto batch = _feature_table.local().encode_version_fence(
+          cluster::cluster_version{9});
```
nit: could you add a comment about what's special about cluster version 9? Alternatively maybe we could alias version 9 as some `earliest_version_with_offset_retention` or something

andrwng commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -746,6 +762,75 @@ ss::future<> group_manager::do_recover_group(
+                  "Cannot write offset retention version fence for partition "
+                  "{}: not leader",
+                  p->partition->ntp());
+                break;
```
nit: here and elsewhere `co_return` better signals the intent to readers IMO (especially since this is a kinda tall method)

dotnwat replied:
@andrwng thanks I posted a PR addressing your nits: https://github.com/redpanda-data/redpanda/pull/8660

andrwng reviewed (approved):
LGTM, just nits, nothing blocking
